yt/yt/server/lib/io/io_engine_fair.cpp

```diff
@@ -0,0 +1,841 @@
+        for (auto c : callbacks) {
+            newCallbacks.push_back({c, ExpectedBytes_});
+        }
+        Underlying_->InvokeWithExpectedBytes(newCallbacks);
+    }
+
+    //! Returns the thread id this invoker is bound to.
+    //! For invokers not bound to any particular thread,
+    //! returns |InvalidThreadId|.
+    NThreading::TThreadId GetThreadId() const override {
+        return Underlying_->GetThreadId();
+    }
+
+    //! Returns true if this invoker is either equal to #invoker or wraps it,
+    //! in some sense.
+    bool CheckAffinity(const IInvokerPtr& invoker) const override {
+        return Underlying_->CheckAffinity(invoker);
+    }
+
+    //! Returns true if invoker is serialized, i.e. never executes
+    //! two callbacks concurrently.
+    bool IsSerialized() const override {
+        return Underlying_->IsSerialized();
+    }
+
+    using TWaitTimeObserver = std::function<void(TDuration)>;
+    void RegisterWaitTimeObserver(TWaitTimeObserver waitTimeObserver) override {
+        return Underlying_->RegisterWaitTimeObserver(waitTimeObserver);
+    }
+
+private:
+    IInvokerWithExpectedBytesPtr Underlying_;
+    i64 ExpectedBytes_;
+};
+
+DEFINE_REFCOUNTED_TYPE(TWrappedInvokerWithExpectedBytes)
+
+template <typename TRequestSlicer>
+class TFairIOEngine
+    : public TIOEngineBase
+{
+public:
+    using TConfig = TFairIOEngineConfig;
+    using TConfigPtr = TIntrusivePtr<TConfig>;
+
+    TFairIOEngine(
+        TFairIOEngineConfigPtr config,
+        TString locationId,
+        TProfiler profiler,
+        NLogging::TLogger logger)
+        : TIOEngineBase(
+            config,
+            std::move(locationId),
+            std::move(profiler),
+            std::move(logger))
+        , StaticConfig_(std::move(config))
+        , ThreadPool_(StaticConfig_, LocationId_, Logger)
+        , Config_(StaticConfig_)
+        , RequestSlicer_(StaticConfig_->DesiredRequestSize, StaticConfig_->MinRequestSize)
+    { }
+
+    TFuture<TReadResponse> Read(
+        std::vector<TReadRequest> requests,
+        const TWorkloadDescriptor& descriptor,
+        TRefCountedTypeCookie tagCookie,
+        const TSessionId& sessionId,
+        bool useDedicatedAllocations) override
+    {
+        std::vector<TFuture<void>> futures;
+        futures.reserve(requests.size());
+
+        auto invoker = ThreadPool_.GetInvoker(descriptor, sessionId);
+
+        i64 paddedBytes = 0;
+        for (const auto& request : requests) {
+            paddedBytes += GetPaddedSize(request.Offset, request.Size, DefaultPageSize);
+        }
+        auto buffers = AllocateReadBuffers(requests, tagCookie, useDedicatedAllocations);
+
+        for (int index = 0; index < std::ssize(requests); ++index) {
+            for (auto& slice : RequestSlicer_.Slice(std::move(requests[index]), buffers[index])) {
+                i64 expectedBytes = Config_.Acquire()->AdditiveCostOfOperationInBytes + slice.Request.Size;
```

Comment:
Described in a comment in config
```
    // Fair IO engine tries to estimate disk load of a single operation by the number of bytes submitted to disk.
    // This approximation is bad for HDDs because even small request can force HDD to rotate and take some time.
    // AdditiveCostOfOperationInBytes increases estimation for all reqests by this amount of bytes.
    // This applies only to read and write operations.
    int AdditiveCostOfOperationInBytes;
```